### PR TITLE
Describes how TLS and certificates are used by agents.

### PR DESCRIPTION
**Pull request by mfoltzgoogle**
This addresses the following issues:

* Issue #135: Propose TLS 1.3 certificate types and extensions for key exchange
* Issue #122: Challenge/Response over TLS
* Issue #130: Review attack and mitigation considerations for TLS 1.3

This PR defines what ciphers, signing algorithms, and key types are to be used with TLS.

It also defines the terms `agent certificate` and `agent fingerprint` to clarify how the fingerprint value is linked to the certificate.

A note about TLS extensions:

At the F2F we resolved to not require any extensions [1].  However, the supported_versions, signature_algorithms, supported_groups and key_share extensions are required for conforming implementations of TLS 1.3 [2] and have information necessary for EC key exchange to work at all.

The server_name and cookies extensions are also required for conforming implementations.  Even if they are not useful for our purposes, TLS 1.3 implementations may not work at all of they are not included in the handshake, so it seems safer to require them.

[1] https://www.w3.org/2019/05/23-webscreens-minutes.html#x20
[2] https://tools.ietf.org/html/rfc8446#section-9.2

***
<a href="https://pr-preview.s3.amazonaws.com/webscreens/openscreenprotocol/pull/212.html" title="Last updated on Sep 11, 2019, 3:32 PM UTC (71ffda1)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webscreens/openscreenprotocol/212/0c91480...71ffda1.html" title="Last updated on Sep 11, 2019, 3:32 PM UTC (71ffda1)">Diff</a>

### CODE REVIEW

**Comment by mfoltzgoogle** [comment]
PTAL @pthatcherg, also @nigelcearnshaw may be interested in taking a look.

File: index.bs [hunk 1]
```diff
@@ -346,6 +346,136 @@ connection IDs are chosen, agents are restricted from changing IP or port
...
+value for the `O` key for user interface and debugging purposes. It may use the
+agent implementer's or device manufacturer's location as the value for the
+location keys (`L`, `ST`, and `C`) for user interface and debugging purposes.
+
+If an OSP agent sees an [=agent certificate=] it has not yet verified through
+[[#authentication]], it must treat that agent as unverified and initiate
+authentication with that agent before allowing additional messages to be
+exchanged with that agent (apart from the messages described in [[#metadata]]).
+
+if an OSP agent sees a valid [=agent certificate=] it has verified through
```
Comment (pthatcherg) [inline index.bs]: If

File: index.bs [hunk 2]
```diff
@@ -346,6 +346,136 @@ connection IDs are chosen, agents are restricted from changing IP or port
...
+TLS 1.3 {#tls-13}
+-----------------
+
+When an [=OSP Agent=] makes a QUIC connection to another agent, it must use
+[[RFC8446|TLS 1.3]] to secure the connection.  TLS 1.3 must be used in the following
+application profile:
+
+* Only TLS 1.3 is supported.  Older revisions of TLS 1.2 and SSL are not allowed
+    and Open Screen connections must not be allowed to downgrade to earlier
+    versions.
```
[the remark this answers is not in the document]
Comment (mfoltzgoogle) [reply index.bs]: No.  In theory someone could add support for older TLS to QUIC.  But it seems unlikely.  I removed this bullet as it's redundant with the text above anyway.

File: index.bs [hunk 3]
```diff
@@ -346,6 +346,136 @@ connection IDs are chosen, agents are restricted from changing IP or port
...
+        extension.  `ecdsa_secp384r1_sha384` and `ecdsa_secp512r1_sha512` must be
+        included if they are supported by the agent.
+    * `secp256r1` must be included in the `supported_groups` extension.
+        `secp384r1` and `secp512r1` must be included if they are supported by the
+        agent.
+    * The `key_share` extension must be set to the cryptographic parameters for
+        the negotiated signature algoirithm.
+* The `server_name` extension should be set to the following `host_name`:
+    `<fp>._openscreen._udp.local`.
+    * `<fp>` should be substituted with the [=agent fingerprint=].
```
[the remark this answers is not in the document]
Comment (mfoltzgoogle) [reply index.bs]: "agent fingerprint" is defined to have the serialization in mDNS TXT.  But I added a note to make this clear.

File: index.bs [hunk 4]
```diff
@@ -346,6 +346,136 @@ connection IDs are chosen, agents are restricted from changing IP or port
...
+
+When an [=OSP Agent=] makes a QUIC connection to another agent, it must use
+[[RFC8446|TLS 1.3]] to secure the connection.  TLS 1.3 must be used in the following
+application profile:
+
+* Only TLS 1.3 is supported.  Older revisions of TLS 1.2 and SSL are not allowed
+    and Open Screen connections must not be allowed to downgrade to earlier
+    versions.
+    * The `supported_versions` extension must include only `0x0304`.
+* Session resumption with pre-shared-keys is not allowed.
```
[the remark this answers is not in the document]
Comment (mfoltzgoogle) [reply index.bs]: Hm.  The notes from Berlin are somewhat inconclusive about whether to allow session resumption or not.  We should figure that out and then decide about 0-RTT data.

Removed from this PR and added to the issue below.

File: index.bs [hunk 5]
```diff
@@ -346,6 +346,136 @@ connection IDs are chosen, agents are restricted from changing IP or port
...
+        included if they are supported by the agent.
+    * `secp256r1` must be included in the `supported_groups` extension.
+        `secp384r1` and `secp512r1` must be included if they are supported by the
+        agent.
+    * The `key_share` extension must be set to the cryptographic parameters for
+        the negotiated signature algoirithm.
+* The `server_name` extension should be set to the following `host_name`:
+    `<fp>._openscreen._udp.local`.
+    * `<fp>` should be substituted with the [=agent fingerprint=].
+* 0-RTT data is not allowed.
```
[the remark this answers is not in the document]
Comment (mfoltzgoogle) [reply index.bs]: Removed.

File: index.bs [hunk 6]
```diff
@@ -346,6 +346,136 @@ connection IDs are chosen, agents are restricted from changing IP or port
...
+* All other TLS 1.3 extensions must be ignored by OSP agents.
+
+Note: As this is an application-specific profile for TLS 1.3, it is not required
+for OSP agents to interoperate with non-OSP TLS 1.3 endpoints.
+
+Issue: Determine if there advantage to session resumption outside of early data.
```
[the remark this answers is not in the document]
Comment (mfoltzgoogle) [reply index.bs]: Done.

File: index.bs [hunk 7]
```diff
@@ -346,6 +346,136 @@ connection IDs are chosen, agents are restricted from changing IP or port
...
+* Self-signed
+* Supporting the at least one of the following signature algorithms:
+    * secp256r1_sha256
+    * secp384r1_sha384
+    * secp521r1_sha512
+* Valid for signing
+
+The following X.509 v3 fields are to be set as follows:
+
+m<table>
```
[the remark this answers is not in the document]
Comment (mfoltzgoogle) [reply index.bs]: Fixed.

File: index.bs [hunk 8]
```diff
@@ -346,6 +346,136 @@ connection IDs are chosen, agents are restricted from changing IP or port
...
+   <td>Certificate Key usage</td>
+   <td>Signing</td>
+ </tr>
+</tbody>
+</table>
+
+Mandatory fields not mentioned above should be set according to [[RFC 5280]].
+
+The value `<fp>` above should be substituted with the [=agent fingerprint=].
+
```
[the remark this answers is not in the document]
Comment (mfoltzgoogle) [reply index.bs]: Also clarified here.

File: index.bs [hunk 9]
```diff
@@ -346,6 +346,136 @@ connection IDs are chosen, agents are restricted from changing IP or port
...
+Note: The OSP agent may use the agent implementer's or device model name as the
```
[the remark this answers is not in the document]
Comment (mfoltzgoogle) [reply index.bs]: Fixed.

File: index.bs (hunk 1 above)
Comment (mfoltzgoogle) [reply index.bs]: Done

**Comment by mfoltzgoogle** [comment]
Addressed review, PTAL

**Comment by nigelcearnshaw** [comment]
A comment on key_share TLS extension.
"The key_share extension must be set to the cryptographic parameters for the negotiated signature algoirithm"

Note that Key_share is associated with the group and public keys by which the parties derive their shared keys for traffic encryption etc, not the signature public keys (which are in the certificates)

This seems to be  saying that both the signature and DH key exchange parameters must use the same groups? I can't see why that should be the case. Even if they were, the ephemeral keys should be fresh.

More typically this should be (from rfc8446)

The key_share extension must be set to the cryptographic parameters. 

We could more helpfully say

The key_share extension must be set to the cryptographic parameters for the TLS key exchange as per RFC 8446 Section 4.2.8 consistent with the supported_groups extension.

**Pull request merged**